tool/tctl/common/devices.go

```diff
@@ -0,0 +1,378 @@
+	lock   deviceLockCommand
+}
+
+type osType = string
+
+const (
+	linuxType   osType = "linux"
+	macosType   osType = "macos"
+	windowsType osType = "windows"
+)
+
+var osTypes = []string{linuxType, macosType, windowsType}
+
+var osTypeToEnum = map[osType]devicepb.OSType{
+	linuxType:   devicepb.OSType_OS_TYPE_LINUX,
+	macosType:   devicepb.OSType_OS_TYPE_MACOS,
+	windowsType: devicepb.OSType_OS_TYPE_WINDOWS,
+}
+
+func (c *DevicesCommand) Initialize(app *kingpin.Application, cfg *servicecfg.Config) {
+	devicesCmd := app.Command("devices", "Register and manage trusted devices").Hidden()
+
+	addCmd := devicesCmd.Command("add", "Register managed devices.")
+	addCmd.Flag("os", "Operating system").
+		Required().
+		EnumVar(&c.add.os, osTypes...)
+	addCmd.Flag("asset-tag", "Inventory identifier for the device (e.g., Mac serial number)").
+		Required().
+		StringVar(&c.add.assetTag)
+	addCmd.Flag("enroll", "If set, creates a device enrollment token").
+		BoolVar(&c.add.enroll)
+	addCmd.Flag("enroll-ttl", "Time duration for the enrollment token").
+		DurationVar(&c.add.enrollTTL)
+
+	_ = devicesCmd.Command("ls", "Lists managed devices.")
+
+	rmCmd := devicesCmd.Command("rm", "Removes a managed device.")
+	rmCmd.Flag("device-id", "Device identifier").StringVar(&c.rm.deviceID)
+	rmCmd.Flag("asset-tag", "Inventory identifier for the device").StringVar(&c.rm.assetTag)
+
+	enrollCmd := devicesCmd.Command("enroll", "Creates a new device enrollment token.")
+	enrollCmd.Flag("device-id", "Device identifier").StringVar(&c.enroll.deviceID)
+	enrollCmd.Flag("asset-tag", "Inventory identifier for the device").StringVar(&c.enroll.assetTag)
+	enrollCmd.Flag("ttl", "Time duration for the enrollment token").DurationVar(&c.enroll.ttl)
+
+	lockCmd := devicesCmd.Command("lock", "Locks a device.")
+	lockCmd.Flag("device-id", "Device identifier").StringVar(&c.lock.deviceID)
+	lockCmd.Flag("asset-tag", "Inventory identifier for the device").StringVar(&c.lock.assetTag)
+	lockCmd.Flag("message", "Message to display to locked-out users").StringVar(&c.lock.message)
+	lockCmd.Flag("expires", "Time point (RFC3339) when the lock expires").StringVar(&c.lock.expires)
+	lockCmd.Flag("ttl", "Time duration after which the lock expires").DurationVar(&c.lock.ttl)
+}
+
+// runner is used as a simple interface for subcommands.
+type runner interface {
+	Run(context.Context, auth.ClientI) error
+}
+
+func (c *DevicesCommand) TryRun(ctx context.Context, selectedCommand string, authClient auth.ClientI) (match bool, err error) {
+	innerCmd, ok := map[string]runner{
+		"devices add":    &c.add,
+		"devices ls":     &c.ls,
+		"devices rm":     &c.rm,
+		"devices enroll": &c.enroll,
+		"devices lock":   &c.lock,
+	}[selectedCommand]
+	if !ok {
+		return false, nil
+	}
+
+	switch err := trail.FromGRPC(innerCmd.Run(ctx, authClient)); {
+	case trace.IsNotImplemented(err):
+		return true, trace.AccessDenied("Device Trust requires a Teleport Enterprise Auth Server running v12 or later.")
```

Comment:
This error handling is the only part that's not verbatim copied from the current implementation in e.